feat(inputs.x509_cert): Add OCSP stapling information for leaf certificates (#10550)

jjh74 opened the pull request:
# Required for all PRs

<!-- Before opening a pull request you should run the following checks to make sure the CI will pass.

make check
make check-deps
make test
make docs

-->

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [ ] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

resolves #10550

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

example: Fixed an off by one error in counter variable in type FooBar.

example: Added an input plugin to gather yak shaving metrics using
golang library yaktech/shaver. -->

This PR adds OCSP stapling information from tls connectionstate(https://pkg.go.dev/crypto/tls#ConnectionState) (https://pkg.go.dev/golang.org/x/crypto/ocsp#ParseResponse). OCSPResponse signature is not checked against issuer certificate.

New fields: ocsp_status(Good=0,Revoked=1,Unknown=2), ocsp_produced_at, ocsp_this_update, ocsp_next_update and ocsp_revoked_at (when ocsp_status=2).
New tags: ocsp_stapled yes/no

srebhan commented:
@jjh74 can you please run `make tidy` to make CI happy!?

srebhan commented on plugins/inputs/x509_cert/x509_cert.go:
```diff
@@ -101,52 +106,75 @@ func (c *X509Cert) Gather(acc telegraf.Accumulator) error {
+						fields["ocsp_error"] = err.Error()
+					} else {
+						tags["ocsp_stapled"] = "yes"
+						// resp.Status: 0=Good 1=Revoked 2=Unknown
+						fields["ocsp_status"] = resp.Status
```
Can we please make this a tag with the "human-readable" text?

jjh74 replied:
Like verification=valid tag and verification_code=0i field.
Would tag: ocsp_status=good field: ocsp_status_code=0i be ok ?
(I'd like to keep field, I think it'll probably make alert rules easier (at least with prometheus)).

srebhan replied:
Yes that's what I had in mind.

srebhan commented on plugins/inputs/x509_cert/x509_cert.go:
```diff
@@ -186,7 +214,7 @@ func (c *X509Cert) serverName(u *url.URL) (string, error) {
 }
 
-func (c *X509Cert) getCert(u *url.URL, timeout time.Duration) ([]*x509.Certificate, error) {
+func (c *X509Cert) getCert(u *url.URL, timeout time.Duration) (*X509CertificatesOcsp, error) {
```
Can we please just return the OCSP reponse as another return item instead of constructing/deconstructing that new struct? If possible, even parse it here into the expected struct...

srebhan reviewed (changes requested):
@jjh74 thanks for looking into this! I have one concern regarding the assumed order of certificates in the answer returned by the server. Currently, we assume (and also you do it) that we _always_ receive the leaf certificate as the first entry. However, I do not see any guarantee for that order, neither by RFC2459 nor by RFC5280 nor by the Golang code.
So IMO we need to order the certs, i.e. construct the trust-path(s)/tree and then determine the leaf nodes instead of blindly use index zero...

What do you think?

jjh74 commented on plugins/inputs/x509_cert/x509_cert.go:
```diff
@@ -101,52 +106,75 @@ func (c *X509Cert) Gather(acc telegraf.Accumulator) error {
+					fields["verification_code"] = 1
+					fields["verification_error"] = err.Error()
+				}
+				// OCSPResponse only for leaf cert
+				if i == 0 && len(certs.OCSPResponse) > 0 {
```
(exclude_root_certs config option also assumes leaf cert is at index 0).

https://pkg.go.dev/crypto/tls#ConnectionState PeerCertificates/VerifiedChains documentation is not very clear. It does say "_in the order in which they were sent_", but also says "_The first element is the leaf certificate that the connection is verified against_"
And example in https://pkg.go.dev/crypto/tls#example-Config-VerifyConnection seems to assume leaf is PeerCertificates[0].

Maybe PeerCertificates[0] is leaf when `InsecureSkipVerify` is false ? Would it be possible to let tls.Client do the verification (for tcp/https) or does the handshake fail if server cert can't be verfied ?

I quickly tested with apache/mod_ssl and haproxy, and both won't start if the leaf certificate is not first (AH02565: Certificate and private key ... do not match. / inconsistencies between private key and certificate loaded). So probably leaf at index=0 is more common.

jjh74 replied:
Looking at golang code (https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/crypto/tls/handshake_client.go;l=515-532;drc=0301c6c3512561b85b48d0e167f3e405484f496f and https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/crypto/tls/handshake_client.go;drc=0301c6c3512561b85b48d0e167f3e405484f496f;bpv=1;bpt=1;l=851?gsn=verifyServerCertificate&gs=kythe%3A%2F%2Fgo.googlesource.com%2Fgo%3Flang%3Dgo%3Fpath%3Dcrypto%2Ftls%23method%2520Conn.verifyServerCertificate) it kind of looks like that the code does use certificate at index=0 as server/leaf cert ?

(Looks like tls.Client will fail handshake if verification fails, so _"let tls.Client do the verification (for tcp/https)"_ won't work).

(Nginx also seems to require leaf certificate first (http://nginx.org/en/docs/http/configuring_https_servers.html#chains)).

jjh74 replied:
[RFC8446](https://www.rfc-editor.org/rfc/rfc8446)(tls1.3) (4.4.2): "_The sender's certificate MUST come in the first CertificateEntry in the list_"
[RFC5246](https://www.rfc-editor.org/rfc/rfc5246)(tls1.2) (7.4.2): "_The sender's certificate MUST come first in the list._".

I'm not sure if these chapters are correct ones for how server sends cert+chain, but maybe assuming that leaf/server cert is at index=0 is reasonable ?

srebhan replied:
Thanks for looking into this!!! Much appreciated. Given this evidence, I think it is save to assume that the leaf cert comes first.

srebhan commented on go.mod:
```diff
@@ -193,6 +193,19 @@ require (
 	modernc.org/sqlite v1.19.2
 )
 
+require (
+	github.com/Azure/azure-storage-blob-go v0.15.0 // indirect
+	github.com/apache/arrow/go/arrow v0.0.0-20211112161151-bc219186db40 // indirect
+	github.com/aws/aws-sdk-go-v2/feature/s3/manager v1.11.33 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/v4a v1.0.14 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/checksum v1.1.18 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/s3shared v1.13.17 // indirect
+	github.com/aws/aws-sdk-go-v2/service/s3 v1.27.11 // indirect
+	github.com/form3tech-oss/jwt-go v3.2.5+incompatible // indirect
+	github.com/gabriel-vasile/mimetype v1.4.1 // indirect
+	github.com/google/flatbuffers v2.0.8+incompatible // indirect
+)
+
 require (
```
Can you please merge those two sections and run `make tidy`.

srebhan commented on plugins/inputs/x509_cert/x509_cert.go:
```diff
@@ -11,6 +11,7 @@ import (
 	"encoding/pem"
 	"errors"
 	"fmt"
+	"golang.org/x/crypto/ocsp"
```
Please move this down to the group of 3rd-party libs.

srebhan reviewed (commented):
Nice update @jjh74! Only two small comments left. Please also resolve the merge conflict...

jjh74 commented:
> @jjh74 any news on this PR?

I tried to do the go.mod require merge + moved `import "golang.org/x/crypto/ocsp"`. I hope I got it right.
(circleci complains about make check-deps (LICENSE_OF_DEPENDENCIES.md), is this something I should try to fix ?)

I realized that `ocsp.ParseResponse(*ocspresp, nil)` doesn't verify the ocsp stapling response signature against issuer certificate (https://pkg.go.dev/golang.org/x/crypto/ocsp#ParseResponse). I can try to see if it's possible to figure out issuer cert and verify ocsp stapling response or if this PR looks good now I can work on ocsp response verify on different PR ?

srebhan commented:
Yeah I noticed that. It seems like you are adding those new libs as dependencies. Maybe try to rebase on latest master to fix the merge conflicts and see if the issue is gone. If not, you need to add the two entries to the `docs/LICENSE_OF_DEPENDENCIES.md` file...

jjh74 commented:
> Yeah I noticed that. It seems like you are adding those new libs as dependencies. Maybe try to rebase on latest master to fix the merge conflicts and see if the issue is gone. If not, you need to add the two entries to the `docs/LICENSE_OF_DEPENDENCIES.md` file...

I rebased to master and that seemed to work.

After rebase I added two more commits:
1. If server sends certificate chain, try to figure out issuer certificate and verify ocsp stapling response with issuer cert. If verification fails retry `ocsp.ParseResponse` w/out issuer cert.
2. Updated README.md with ocsp tags/fields (run-readme-linter seems to fail (but AFAIK on different plugin/README))

powersj commented:
@jjh74 I pushed a few commits to your PR. Namely, I rebased on master and fixed the conflicts and one linter issue. This should hopefully get us nearly there. I can do a review in a bit.

srebhan reviewed (approved):
Looks good to me. Thanks @jjh74 for the contribution and thanks @powersj for driving this over the finish line.

xoxys commented:
Thanks everyone for the implementation!
